### ISSUE TITLE
Data driven 2.0

### ISSUE THREAD

**Issue by raniejade** [JetBrains/spek#423]
Introduce `where` that can be used to create data-driven tests. 

```kotlin
object MyDataDrivenSpec: Spek({
    val x by memoized<Int>()
    val y by memoized<Int>()
    val expected by memoized<Int>()
    
    test("$x + $y = $expected") {
        assertEqual(expected, x + y)
    }
    
    where(
        c("x",  "y", "expected"),
        r( 1 ,   2 ,         3 ),
        r( 2 ,   2 ,         4 ),
        r( 1 ,  -1 ,         0 ),
    )
})
```

Implementation wise it can be built on top of #411, repeating the scope `n` times and creating pre-defined memoized values based on the declared columns/values.

//cc @ming13 @artem-zinnatullin @hhariri

**Comment by artem-zinnatullin** [comment]
hmm, so it's reflection based and not type-safe

Also how would Spek know that `test(…)` doesn't need to be executed on its own?

**Comment by arturdryomov** [comment]
I think we can promote data-driven approach using patterns instead of implementation.

Simple scenario:

```kotlin
listOf("text A", "text B", "text C").forEach { text ->

    context("sending text [$text]") {

        beforeEachTest {
            service.send(text)
        }

        it("shows progress") {
            verify(view).showProgress()
        }
    }
}
```

Complex scenario:

```kotlin
data class Arguments(val x: Int, val y: Int)

listOf(Arguments(2, 2), Arguments(0, 0)).forEach { arguments ->

    context("sending arguments [$arguments]") {

        beforeEachTest {
            service.send(arguments)
        }

        it("shows progress") {
            verify(view).showProgress()
        }
    }
}
```

It is simple and language-driven. It is possible to do custom formatting, filter values using `if` and just go crazy.

With the data-driven extension we saw the limitation of the centralized _helpers_ since it did not fit custom behaviors, like having 100 arguments or whatever. The reason is quite simple — we are trying to predict test-level code structure from the perspective of a test framework and of course failing to do so. Leveraging language features avoids doing so, at the same time achieving the goal of iterating over values. And we don’t have to maintain extra features as a bonus.

**Comment by Rosomack** [comment]
I agree with @ming13 on this. However I've got a few comments on the usability of @ming13 's proposed solution I'd like to make you aware of. I've tested this on `1.1.5`, apologies if this has already been fixed in `2.0`. 

For complex test cases requiring a few setup parameters, I would expect to be able to break them up into `groups` to improve readability in the IDE like so:
```
    arrayOf(Triple(false, false, false),
        Triple(false, true, true),
        Triple(true, false, true),
        Triple(true, true, false)).forEach {
        given("x equals ${it.first}") {
            given("y equals ${it.second}") {
                on("xoring x and y") {
                    val result = it.first.xor(it.second)
                    it("outputs ${it.third}") {
                        assert(result).isEqualTo(it.third)
                    }
                }
            }
        }
    }
```
This is overengineered on purpose, to illustrate a problem. I believe spek ignores groups that are already discovered under the same name. To put it more simply:

```
    given("A test") {
        on("1 + 2") {
            val result = 1 + 2
            it("equals 3") {
                assert(result).isEqualTo(3)
            }
        }
    }

    given("A test") {
        on("2 + 3") {
            val result = 2 + 3
            it("equals 5") {
                assert(result).isEqualTo(5)
            }
        }
    }

    given("A test") {
        on("3 + 5") {
            val result = 3 + 5
            it("equals 8") {
                assert(result).isEqualTo(8)
            }
        }
    }
```
In the above example, only the first test group will be executed, one test of three.
 
While this does not really hurt much if you're writing test cases by hand, because you don't tend to have groups with duplicate descriptions on the same nesting level, in generated tests it's a problem. I believe duplicate groups should be allowed if technically possible, as a last resort explicitly disallowed by crashing in the discovery phase. Silently ignoring them is not the best thing to do.

**Comment by arturdryomov** [comment]
@Rosomack, hey, that was one of the first Spek issues I’ve opened. See #210. Probably should be fixed in `2.x`.
